File: shoring/blackjackEngine/src/main/java/com/panopset/blackjackEngine/StrategyAction.java
package com.panopset.blackjackEngine;

import static com.panopset.blackjackEngine.CommandDefinitions.CMD_DOUBLE;
import static com.panopset.blackjackEngine.CommandDefinitions.CMD_HIT;
import static com.panopset.blackjackEngine.CommandDefinitions.CMD_SPLIT;
import static com.panopset.blackjackEngine.CommandDefinitions.CMD_STAND;
import static com.panopset.blackjackEngine.CommandDefinitions.CMD_SURRENDER;

public class StrategyAction {

  private final HandPlayer hand;
  private final boolean canDouble;
  private final CasinoRules casinoRules;
  private final boolean singleDeck;

  public StrategyAction(Situation situation, CasinoRules casinoRules) {
    this.hand = situation.getHandPlayer();
    this.canDouble = hand.canDouble();
    this.casinoRules = casinoRules;
    this.singleDeck = casinoRules.decks == 1;
  }

  public String getRecommendedAction(String recommendedActionText) {
    switch (recommendedActionText) {
      case "H":
        return CMD_HIT;
      case "P":
        return CMD_SPLIT;
      case "H*":
        return getHstarRecommendation();
      case "Hx":
        return getHxRecommendation();
      case "Hd":
        return getHdRecommendation();
      case "H@":
        return getHatRecommenation();
      case "HT":
        return getHtRecommendation();
      case "Rs":
        return getRsRecommendation();
      case "Rh":
        return getRhRecommendation();
      case "Sh":
        return getShRecommendation();
      case "Dh":
        return getDhRecommendation();
      case "Ds":
        return getDsRecommendation();
      default:
        return CMD_STAND;
    }
  }

  private String getHstarRecommendation() {
    if (singleDeck) {
      return getDhRecommendation();
    }
    return CMD_HIT;
  }

  private String getHxRecommendation() {
    if (singleDeck) {
      return getDsRecommendation();
    }
    return CMD_HIT;
  }

  private String getHatRecommenation() {
    if (singleDeck) {
      return CMD_SPLIT;
    }
    return CMD_HIT;
  }

  private String getHtRecommendation() {
    if (hand.isInitialDeal()) {
      for (BlackjackCard card : hand.cards) {
        if (card.getHardValue() == 10) {
          return CMD_HIT;
        }
      }
    }
    return CMD_STAND;
  }

  private String getRsRecommendation() {
    if (casinoRules.isLateSurrenderAllowed) {
      return CMD_SURRENDER;
    }
    return CMD_STAND;
  }

  private String getRhRecommendation() {
    if (casinoRules.isLateSurrenderAllowed) {
      return CMD_SURRENDER;
    }
    return CMD_HIT;
  }

  private String getShRecommendation() {
    for (BlackjackCard card : hand.cards) {
      if (card.getHardValue() == 4 || card.getHardValue() == 5) {
        return CMD_STAND;
      }
    }
    return CMD_HIT;
  }

  private String getHdRecommendation() {
    if (canDouble && (casinoRules.dealerHitsDoft17 || casinoRules.decks < 3)) {
      return CMD_DOUBLE;
    }
    return CMD_HIT;
  }

  private String getDsRecommendation() {
    return canDouble ? CMD_DOUBLE : CMD_STAND;
  }

  private String getDhRecommendation() {
    return canDouble ? CMD_DOUBLE : CMD_HIT;
  }
}

File: code/shoring/blackjackEngine/src/main/kotlin/com/panopset/blackjackEngine/StrategyAction.kt
package com.panopset.blackjackEngine

class StrategyAction(situation: Situation, private val casinoRules: CasinoRules) {
    private val hand: HandPlayer?
    private val canDouble: Boolean
    private val singleDeck: Boolean

    init {
        hand = situation.handPlayer
        canDouble = hand?.canDouble(casinoRules.isDoubleAfterSplitAllowed) ?: false
        singleDeck = casinoRules.decks == 1
    }

    fun getRecommendedAction(recommendedActionText: String?): String {
        return when (recommendedActionText) {
            "H" -> CMD_HIT
            "P" -> CMD_SPLIT
            "H*" -> hstarRecommendation
            "Hx" -> hxRecommendation
            "Hd" -> hdRecommendation
            "H@" -> hatRecommenation
            "HT" -> htRecommendation
            "Rs" -> rsRecommendation
            "Rh" -> rhRecommendation
            "Sh" -> shRecommendation
            "Dh" -> dhRecommendation
            "Ds" -> dsRecommendation
            else -> CMD_STAND
        }
    }

    private val hstarRecommendation: String
        get() = if (singleDeck) {
            dhRecommendation
        } else CMD_HIT
    private val hxRecommendation: String
        get() = if (singleDeck) {
            dsRecommendation
        } else CMD_HIT
    private val hatRecommenation: String
        get() = if (singleDeck) {
            CMD_SPLIT
        } else CMD_HIT
    private val htRecommendation: String
        get() {
            if (hand == null) {
                return CMD_STAND
            }
            if (hand.isInitialDeal) {
                for (card in hand.cards) {
                    if (card.hardValue == 10) {
                        return CMD_HIT
                    }
                }
            }
            return CMD_STAND
        }
    private val rsRecommendation: String
        get() = if (casinoRules.isLateSurrenderAllowed) {
            CMD_SURRENDER
        } else CMD_STAND
    private val rhRecommendation: String
        get() = if (casinoRules.isLateSurrenderAllowed) {
            CMD_SURRENDER
        } else CMD_HIT
    private val shRecommendation: String
        get() {
            if (hand == null) {
                return CMD_STAND
            }
            for (card in hand.cards) {
                if (card.hardValue == 4 || card.hardValue == 5) {
                    return CMD_STAND
                }
            }
            return CMD_HIT
        }
    private val hdRecommendation: String
        get() = if (canDouble && (casinoRules.dealerHitsSoft17 || casinoRules.decks < 3)) {
            CMD_DOUBLE
        } else CMD_HIT
    private val dsRecommendation: String
        get() = if (canDouble) CMD_DOUBLE else CMD_STAND
    private val dhRecommendation: String
        get() = if (canDouble) CMD_DOUBLE else CMD_HIT
}
